### PR TITLE
Typecheck support for verifying providerless build

### PR DESCRIPTION
**Pull request by BenTheElder**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
/kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**: Adds support to the typecheck utility for:
- ignoring extra directories
- skipping _test.go code
- specifying build tags

This is necessary to typecheck providerless build. We want to:
- set the providerless tag
- not check unit tests (we only build the binaries in providerless mode to validate provider removal from the core, we're not prepared to handle tests yet)
- not check test/* (e2e etc. do not have an agreed plan for removing in-tree providers yet)

Currently this gives us:
```
$ hack/verify-typecheck.sh --skip-test --tags=providerless --ignore-dirs=test
type-checking:  linux/amd64, windows/386, darwin/amd64, linux/arm, linux/386, windows/amd64, linux/arm64, linux/ppc64le, linux/s390x, darwin/386
ERROR(all) cmd/kube-controller-manager/app/plugins.go:141:70: too many arguments                         
ERROR(all) cmd/kube-controller-manager/app/plugins.go:141:20: cannot assign 1 values to 2 variables
ERROR(all) cmd/kube-controller-manager/app/plugins.go:88:70: too many arguments
ERROR(all) cmd/kube-controller-manager/app/plugins.go:88:20: cannot assign 1 values to 2 variables
ERROR(all) cmd/kube-controller-manager/app/plugins.go:63:70: too many arguments
ERROR(all) cmd/kube-controller-manager/app/plugins.go:63:20: cannot assign 1 values to 2 variables
ERROR(all) cmd/kubelet/app/plugins.go:68:60: too many arguments                                          
ERROR(all) cmd/kubelet/app/plugins.go:68:20: cannot assign 1 values to 2 variables
ERROR(all) pkg/controller/nodeipam/node_ipam_controller.go:135:83: cannot use nodeCIDRMaskSizes (variable of type []int) as int value in argument to startLegacyIPAM
ERROR(darwin) pkg/volume/awsebs/attacher_unsupported.go:23:17: undeclared name: awsElasticBlockStoreAttacher
ERROR(windows) pkg/volume/awsebs/attacher_windows.go:31:17: undeclared name: awsElasticBlockStoreAttacher
ERROR(windows) pkg/volume/awsebs/attacher_windows.go:38:17: undeclared name: awsElasticBlockStoreAttacher
ERROR(windows) pkg/volume/awsebs/attacher_windows.go:43:32: undeclared name: awsElasticBlockStorePluginName
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:25:17: undeclared name: awsElasticBlockStoreAttacher
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:26:17: undeclared name: getDiskByIDPaths
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:26:38: KubernetesVolumeID not declared by package aws
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:27:9: undeclared name: verifyDevicePath
ERROR(windows) pkg/volume/vsphere_volume/vsphere_volume_util_windows.go:36:29: undeclared name: diskByIDPath
ERROR(windows) pkg/volume/vsphere_volume/vsphere_volume_util_windows.go:54:43: undeclared name: diskByIDPath
ERROR(windows) pkg/volume/vsphere_volume/vsphere_volume_util_windows.go:54:56: undeclared name: diskSCSIPrefix
12260/12260 linux/386     translations/test/en_US/LC_MESSAGES                                             
exit status 1
!!! Type Check has failed. This may cause cross platform build failures.
!!! Please see https://git.k8s.io/kubernetes/test/typecheck for more information.
```

XREF: https://github.com/kubernetes/kubernetes/pull/85456

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by BenTheElder** [comment]
Added a verify script for this that we can run in CI.

**Comment by BenTheElder** [comment]
fixed the unit test(s)
/hold cancel

**Comment by BenTheElder** [comment]
```console
$ hack/verify-typecheck-providerless.sh
stat /usr/local/google/home/bentheelder/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/bazel-bin: no such file or directory
type-checking:  linux/amd64, windows/386, darwin/amd64, linux/arm, linux/386, windows/amd64, linux/arm64, linux/ppc64le, linux/s390x, darwin/386
ERROR(all) cmd/kube-controller-manager/app/plugins.go:141:70: too many arguments                          
ERROR(all) cmd/kube-controller-manager/app/plugins.go:141:20: cannot assign 1 values to 2 variables
ERROR(all) cmd/kube-controller-manager/app/plugins.go:88:70: too many arguments
ERROR(all) cmd/kube-controller-manager/app/plugins.go:88:20: cannot assign 1 values to 2 variables
ERROR(all) cmd/kube-controller-manager/app/plugins.go:63:70: too many arguments
ERROR(all) cmd/kube-controller-manager/app/plugins.go:63:20: cannot assign 1 values to 2 variables
ERROR(all) cmd/kubelet/app/plugins.go:68:60: too many arguments                                           
ERROR(all) cmd/kubelet/app/plugins.go:68:20: cannot assign 1 values to 2 variables
ERROR(all) pkg/controller/nodeipam/node_ipam_controller.go:135:83: cannot use nodeCIDRMaskSizes (variable of type []int) as int value in argument to startLegacyIPAM
ERROR(windows) pkg/volume/awsebs/attacher_windows.go:31:17: undeclared name: awsElasticBlockStoreAttacher 
ERROR(windows) pkg/volume/awsebs/attacher_windows.go:38:17: undeclared name: awsElasticBlockStoreAttacher 
ERROR(windows) pkg/volume/awsebs/attacher_windows.go:43:32: undeclared name: awsElasticBlockStorePluginName
ERROR(darwin) pkg/volume/awsebs/attacher_unsupported.go:23:17: undeclared name: awsElasticBlockStoreAttacher
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:25:17: undeclared name: awsElasticBlockStoreAttacher
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:26:17: undeclared name: getDiskByIDPaths
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:26:38: KubernetesVolumeID not declared by package aws
ERROR(linux/386,linux/amd64,linux/arm,linux/arm64,linux/ppc64le,linux/s390x) pkg/volume/awsebs/attacher_linux.go:27:9: undeclared name: verifyDevicePath
ERROR(windows) pkg/volume/vsphere_volume/vsphere_volume_util_windows.go:36:29: undeclared name: diskByIDPath
ERROR(windows) pkg/volume/vsphere_volume/vsphere_volume_util_windows.go:54:43: undeclared name: diskByIDPath
ERROR(windows) pkg/volume/vsphere_volume/vsphere_volume_util_windows.go:54:56: undeclared name: diskSCSIPrefix
22230/22230 linux/amd64   translations/test/en_US/LC_MESSAGES                                             
exit status 1
!!! Type Check has failed. This may cause cross platform build failures.
!!! Please see https://git.k8s.io/kubernetes/test/typecheck for more information.
```

**Comment by dims** [comment]
cc @ibuildthecloud

File: test/typecheck/main.go [hunk 1]
```diff
@@ -41,12 +41,15 @@ import (
...
-	defuses   = flag.Bool("defuse", false, "output defs/uses")
-	serial    = flag.Bool("serial", false, "don't type check platforms in parallel")
+	verbose    = flag.Bool("verbose", false, "print more information")
+	cross      = flag.Bool("cross", true, "build for all platforms")
+	platforms  = flag.String("platform", "", "comma-separated list of platforms to typecheck")
+	timings    = flag.Bool("time", false, "output times taken for each phase")
+	defuses    = flag.Bool("defuse", false, "output defs/uses")
+	serial     = flag.Bool("serial", false, "don't type check platforms in parallel")
+	skipTest   = flag.Bool("skip-test", false, "don't type check test code")
+	tags       = flag.String("tags", "", "comma-separated list of additional build tags to apply")
```
Comment (liggitt) [inline test/typecheck/main.go]: clarify what this is in addition to (presumably os/arch-level build tags?)

Comment (BenTheElder) [reply test/typecheck/main.go]: ACK -- it's in addition to go's defaults, we're not setting any ourselves other than these.

File: test/typecheck/main.go [hunk 2]
```diff
@@ -41,12 +41,15 @@ import (
...
+	ignoreDirs = flag.String("ignore-dirs", "", "comma-separated list of directories to ignore")
```
Comment (liggitt) [inline test/typecheck/main.go]: clarify things specified here are in addition to standard ignored dirs

File: test/typecheck/main.go [hunk 3]
```diff
@@ -62,6 +65,24 @@ var (
...
+	// directories we always ignore
+	standardIgnoreDirs = []string{
+		// Staging code is symlinked from vendor/k8s.io, and uses import
+		// paths as if it were inside of vendor/. It fails typechecking
+		// inside of staging/, but works when typechecked as part of vendor/.
+		"staging",
+		// OS-specific vendor code tends to be imported by OS-specific
+		// packages. We recursively typecheck imported vendored packages for
+		// each OS, but don't typecheck everything for every OS.
+		"vendor",
```
Comment (liggitt) [inline test/typecheck/main.go]: this isn't new, but I was slightly surprised by it... I think this means if we add a main package under vendor or staging, typecheck won't check its import tree. Worth remembering as we try to push more things out of root k/k

File: test/typecheck/main.go (hunk 1 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: I'm adding that this is in addition to the defaults, but I don't think we should document the defaults here as they come from go and could change (currently GOOS and GOARCH)

Comment (BenTheElder) [reply test/typecheck/main.go]: (I think also more that are not as explicitly documented though?)

File: test/typecheck/main.go (hunk 3 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: we still typecheck this IIUC, as in, we typecheck files which then import vendor, which transitiviely checks.

we don't typecheck literally all code in vendor though.

File: test/typecheck/main.go (hunk 2 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: done

File: test/typecheck/main.go [hunk 4]
```diff
@@ -79,6 +100,8 @@ func newAnalyzer(platform string) *analyzer {
 	platSplit := strings.Split(platform, "/")
 	ctx.GOOS, ctx.GOARCH = platSplit[0], platSplit[1]
 	ctx.CgoEnabled = true
+	tagsSplit := strings.Split(*tags, ",")
+	ctx.BuildTags = append(ctx.BuildTags, tagsSplit...)
```
Comment (liggitt) [inline test/typecheck/main.go]: This will append an empty string if no tags are specified, right?

File: test/typecheck/main.go [hunk 5]
```diff
@@ -310,7 +328,9 @@ func main() {
 		args = append(args, ".")
 	}
 
-	c := collector{}
+	c := collector{
+		ignoreDirs: append(strings.Split(*ignoreDirs, ","), standardIgnoreDirs...),
```
Comment (liggitt) [inline test/typecheck/main.go]: same question about appending an empty string of no additional ignore dirs are specified

File: hack/verify-typecheck-providerless.sh [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+
+cd "${KUBE_ROOT}"
+# verify the providerless build
+# https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190729-building-without-in-tree-providers.md
+hack/verify-typecheck.sh --skip-test --tags=providerless --ignore-dirs=test
```
Comment (liggitt) [inline hack/verify-typecheck-providerless.sh]: Did we end up having to skip all test files for now? I thought the test/e2e dir was the main thing we needed to skip

File: test/typecheck/main.go (hunk 5 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: `strings.Split` returns an empty slice in this case

File: hack/verify-typecheck-providerless.sh (hunk 6 above)
Comment (BenTheElder) [reply hack/verify-typecheck-providerless.sh]: yeah, there's also some integration code IIRC...
we have no plan of record for removing these from test yet.
IMO it needs to happen, but I'm not signing myself up just yet, and getting the core binaries to work without them is a start.

we still don't have these things fully removed from the core binaries as well :( (credential provider...)

File: test/typecheck/main.go (hunk 4 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: yes, we should probably not do that, however I think it's harmless, go can't have a build constraint on a nothing tag.

Comment (BenTheElder) [reply test/typecheck/main.go]: er no, because we split first, and it will be the empty slice

Comment (BenTheElder) [reply test/typecheck/main.go]: `strings.Split("", ",")` => `[]`

**Comment by BenTheElder** [comment]
e2e-gce flakes https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/85457/pull-kubernetes-e2e-gce/1202459386178441217
/retest

File: test/typecheck/main.go (hunk 4 above)
Comment (liggitt) [reply test/typecheck/main.go]: Are you sure?

https://play.golang.org/p/UcYudBHOokb

File: hack/verify-typecheck-providerless.sh (hunk 6 above)
Comment (liggitt) [reply hack/verify-typecheck-providerless.sh]: sure, was just curious

File: test/typecheck/main.go (hunk 4 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: bah, compare https://play.golang.org/p/Z4_2bLbJdWq which I tested :|

File: test/typecheck/main.go (hunk 5 above)
Comment (BenTheElder) [reply test/typecheck/main.go]: https://github.com/kubernetes/kubernetes/pull/85457#discussion_r354126681 🤦‍♂

**Comment by BenTheElder** [comment]
> tide Pending — Not mergeable. Retesting 4 jobs.

!?

**Comment by BenTheElder** [comment]
I think the merge robot is broken :( raising to testing-ops

**Comment by BenTheElder** [comment]
now to enable this in test-infra (the merge robot was indeed broken and is running again)

**Comment by dims** [comment]
@BenTheElder did we enable this in test-infra?

**Comment by BenTheElder** [comment]
we set `WHAT=typecheck typecheck-providerless` in `pull-kubernetes-verify-typecheck` so it should be enabled

**Pull request merged**